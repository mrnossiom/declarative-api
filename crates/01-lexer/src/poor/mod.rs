mod cursor;
mod lexer;
mod token;

pub use cursor::Cursor;
pub use token::*;
