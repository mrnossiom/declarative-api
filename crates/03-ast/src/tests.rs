use crate::types::*;
use lexer::{span::Span, symbols::Symbol};

#[allow(clippy::too_many_lines)]
fn _dummy_api() -> Api {
	Api {
		meta: vec![
			Attribute {
				ident: Ident {
					symbol: Symbol::new_static("doc"),
					span: Span { start: 0, end: 0 },
				},
				value: " This Api is a test for the documentation generator of the api".into(),
				span: Span { start: 0, end: 0 },
			},
			Attribute {
				ident: Ident {
					symbol: Symbol::new_static("doc"),
					span: Span { start: 0, end: 0 },
				},
				value: " This is a second line of comment".into(),
				span: Span { start: 0, end: 0 },
			},
			Attribute {
				ident: Ident {
					symbol: Symbol::new_static("doc"),
					span: Span { start: 0, end: 0 },
				},
				value: " This is a third line of comment".into(),
				span: Span { start: 0, end: 0 },
			},
		],
		items: vec![
			Item {
				attrs: vec![],
				kind: ItemKind::Meta(Meta {
					fields: vec![
						MetaField {
							ident: Ident {
								symbol: Symbol::new_static("name"),
								span: Span { start: 0, end: 0 },
							},
							value: MetaFieldKind::Str("Wiro's API".into()),
							span: Span { start: 0, end: 0 },
						},
						MetaField {
							ident: Ident {
								symbol: Symbol::new_static("description"),
								span: Span { start: 0, end: 0 },
							},
							value: MetaFieldKind::Str("This is the API of Wiro".into()),
							span: Span { start: 0, end: 0 },
						},
						MetaField {
							ident: Ident {
								symbol: Symbol::new_static("version"),
								span: Span { start: 0, end: 0 },
							},
							value: MetaFieldKind::Str("1.0.0".into()),
							span: Span { start: 0, end: 0 },
						},
						MetaField {
							ident: Ident {
								symbol: Symbol::new_static("name"),
								span: Span { start: 0, end: 0 },
							},
							value: MetaFieldKind::Vec(vec![
								MetaFieldKind::Str("https://paradigm.lighton.ai/api/v1".into()),
								MetaFieldKind::Str(
									"https://paradigm-preprod.lighton.ai/api/v1".into(),
								),
								MetaFieldKind::Str("https://paradigm-dev.lighton.ai/api/v1".into()),
							]),
							span: Span { start: 0, end: 0 },
						},
					],
				}),
				span: Span { start: 0, end: 0 },
				ident: Ident {
					symbol: Symbol::new_static(""),
					span: Span { start: 0, end: 0 },
				},
			},
			Item {
				attrs: vec![Attribute {
					ident: Ident {
						symbol: Symbol::new_static("doc"),
						span: Span { start: 0, end: 0 },
					},
					value: " Imports the `builder.dapi` file".into(),
					span: Span { start: 0, end: 0 },
				}],
				ident: Ident {
					symbol: Symbol::new_static("builder"),
					span: Span { start: 0, end: 0 },
				},
				kind: ItemKind::Scope(ScopeKind::Unloaded),
				span: Span { start: 0, end: 0 },
			},
			Item {
				attrs: vec![],
				ident: Ident {
					symbol: Symbol::new_static("dashboard"),
					span: Span { start: 0, end: 0 },
				},
				kind: ItemKind::Scope(ScopeKind::Loaded {
					items: vec![Item {
						attrs: vec![],
						ident: Ident {
							symbol: Symbol::new_static("dashboard"),
							span: Span { start: 0, end: 0 },
						},
						kind: ItemKind::Path(Path {
							path: PathKind::String(Ident {
								symbol: Symbol::new_static("dashboard"),
								span: Span { start: 0, end: 0 },
							}),
							items: vec![Item {
								attrs: vec![],
								ident: Ident {
									symbol: Symbol::new_static(""),
									span: Span { start: 0, end: 0 },
								},
								kind: ItemKind::Headers(Headers {
									headers: vec![
										HeaderField {
											attrs: vec![
												Attribute {
													ident: Ident {
														symbol: Symbol::new_static("doc"),
														span: Span { start: 0, end: 0 },
													},
													value: " # Safety".into(),
													span: Span { start: 0, end: 0 },
												},
												Attribute {
													ident: Ident {
														symbol: Symbol::new_static("doc"),
														span: Span { start: 0, end: 0 },
													},
													value: " This is a comment".into(),
													span: Span { start: 0, end: 0 },
												},
												Attribute {
													ident: Ident {
														symbol: Symbol::new_static("doc"),
														span: Span { start: 0, end: 0 },
													},
													value: " This is a second line of comment"
														.into(),
													span: Span { start: 0, end: 0 },
												},
												Attribute {
													ident: Ident {
														symbol: Symbol::new_static("description"),
														span: Span { start: 0, end: 0 },
													},
													value: "The API Key of the User".into(),
													span: Span { start: 0, end: 0 },
												},
												Attribute {
													ident: Ident {
														symbol: Symbol::new_static("prefix"),
														span: Span { start: 0, end: 0 },
													},
													value: "Api-Key".into(),
													span: Span { start: 0, end: 0 },
												},
											],
											ident: Ident {
												symbol: Symbol::new_static("Authorization"),
												span: Span { start: 0, end: 0 },
											},
											span: Span { start: 0, end: 0 },
										},
										HeaderField {
											attrs: vec![Attribute {
												ident: Ident {
													symbol: Symbol::new_static("description"),
													span: Span { start: 0, end: 0 },
												},
												value: "The Model of the User".into(),
												span: Span { start: 0, end: 0 },
											}],
											ident: Ident {
												symbol: Symbol::new_static("X-Model"),
												span: Span { start: 0, end: 0 },
											},
											span: Span { start: 0, end: 0 },
										},
									],
								}),
								span: Span { start: 0, end: 0 },
							}],
						}),
						span: Span { start: 0, end: 0 },
					}],
					span: Span { start: 0, end: 0 },
				}),
				span: Span { start: 0, end: 0 },
			},
		],
		span: Span { start: 0, end: 0 },
	}
}
