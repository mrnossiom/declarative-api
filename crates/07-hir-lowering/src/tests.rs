
use std::{
	str::FromStr,
	vec,
};


use super::*;

#[test]
fn test_1() {
	let v1 = types::Version {
		major: 1,
		minor: 2,
		patch: 3,
	};

	let v2 = types::Version {
		major: 1,
		minor: 2,
		patch: 3,
	};
	assert_eq!(v1, v2);
}

#[test]
fn test_2() {
	let v1 = types::Version {
		major: 1,
		minor: 2,
		patch: 3,
	};
	let v2 = types::Version {
		major: 0,
		minor: 3,
		patch: 4,
	};
	assert!(v1 > v2);
}



#[test]
#[allow(clippy::too_many_lines)]
fn test_3() {
	let input_ast = ast::types::Api {
		metadata: ast::types::ApiMetadata {
			name: Some("Wiro's API".into()),
			version: Some(ast::types::Version {
				major: 1,
				minor: 0,
				patch: 0,
			}),
			urls: vec![
				"https://paradigm.lighton.ai/api/v1".into(),
				"https://paradigm-preprod.lighton.ai/api/v1".into(),
				"https://paradigm-dev.lighton.ai/api/v1".into(),
			],
			comment: Some("Je suis un commentaire de documentation (des métadonnées)".into()),
		},
		data: ast::types::ApiData {
			child_scopes: vec![ast::types::Scope {
				name: "dashboard".into(),
				child_scopes: vec![],
				child_models: vec![ast::types::Model{
					name:"metrics".into(),
					model_body: vec![ast::types::KeyValuePair{
						key:"name".into(),
						type_:ast::types::Type::String,
						description:"The name of the metric".into(),
						parameters:vec![],
						comment:None,
					},ast::types::KeyValuePair{
						key:"email".into(),
						type_:ast::types::Type::String,
						description:String::new(),
						parameters:vec![],
						comment:None,
					},ast::types::KeyValuePair{
						key:"password".into(),
						type_:ast::types::Type::String,
						description:String::new(),
						parameters:vec![],
						comment:None,
					},
					],
				comment:None,
				}],
				child_paths: vec![ast::types::Path {
					name: "dashboard".into(),
					child_scopes: vec![],
					child_paths: vec![ast::types::Path {
						name: "metrics".into(),
						child_paths: vec![],
						comment: None,
						methods: vec![],
						headers: vec![ast::types::KeyValuePair {
							key: "Authorization".into(),
							type_: ast::types::Type::String,
							description: "The authorization key.".into(),
							parameters: vec![],
							comment: None,
						}],
						metadata: vec![],
						parameters: vec![],
						query: vec![],
						child_scopes: vec![ast::types::Scope {
							name: "metrics".into(),
							child_scopes: vec![],
							child_models: vec![],
							child_paths: vec![],
							methods: vec![ast::types::Method {
									method: http::Method::from_str("GET").expect(
												"a method used for testing the ast-lowering crate doesn't exist",
											),
											responses: vec![
												ast::types::Response{
													headers: vec![],
													status_code: 200,
													body:vec![
                                                        
                                                    ast::types::KeyValuePair{
														key:"metrics".into(),
														type_:ast::types::Type::Model("metrics".into()),
														description:String::new(),
														parameters:vec![],
														comment: None,
													},ast::types::KeyValuePair{
														key:"status".into(),
														type_:ast::types::Type::Object(
															vec![
																ast::types::KeyValuePair{
																	key:"message".into(),
																	type_:ast::types::Type::String,
																	description:"The status message itself".into(),
																	parameters:vec![],
																	comment: None,
																},ast::types::KeyValuePair{
																	key:"code".into(),
																	type_:ast::types::Type::Int,
																	description:"The status code".into(),
																	parameters:vec![],
																	comment: None,
																}
															]
														),
														description:"A status message container".into(),
														parameters:vec![],
														comment: None,
													}],
													comment: None
												}
											],
											headers: vec![],
											query: vec![],
											comment: None,
										}],
							parameters: vec![],
							headers: vec![],
							query: vec![],
							comment: None,
						}],
					}],
					methods: vec![],
					headers: vec![],
					parameters: vec![],
					query: vec![],
					metadata: vec![],
					comment: None,
				}],
				methods: vec![],
				parameters: vec![],
				headers: vec![],
				query: vec![],
				comment: None,
			}],
		},
	};
	
	
	let mut parsed_intermediate_representation = types::IntermediateRepresentation::default();
	input_ast.parse(
		types::ParserVariables::default(),
		&mut parsed_intermediate_representation,
	);
	let expected_intermediate_representation = utils::gen_test_ir();
	//dbg!(&expected_intermediate_representation);
	//dbg!(&parsed_intermediate_representation);
	assert!(expected_intermediate_representation == parsed_intermediate_representation);
}
