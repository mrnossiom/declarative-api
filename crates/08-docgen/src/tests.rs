use super::*;
use hir_lowering::{utils, Parse};

#[test]
fn test_1() {
	let test_ir = utils::gen_test_ir();
	let mut generated_file_list = HashMap::new();
	test_ir.markdown(&mut generated_file_list);
	// generated_file_list.write(Path::new("test-artifacts/test-1"));
}

#[test]
#[allow(clippy::too_many_lines)]
fn test_2() {
	let input_hir = hir::types::Api {
		metadata: hir::types::ApiMetadata {
			name: Some("Wiro's API".into()),
			version: Some(hir::types::Version {
				major: 1,
				minor: 0,
				patch: 0,
			}),
			urls: vec![
				"https://paradigm.lighton.ai/api/v1".into(),
				"https://paradigm-preprod.lighton.ai/api/v1".into(),
				"https://paradigm-dev.lighton.ai/api/v1".into(),
			],
			comment: Some("This is the API of Wiro".into()),
		},
		data: hir::types::ApiData {
			scopes: vec![hir::types::Scope {
				name: "dashboard".into(),
				models: vec![],
				scopes: vec![],
				methods: vec![],
				paths: vec![hir::types::Path {
					name: "dashboard".into(),
					scopes: vec![],
					paths: vec![],
					methods: vec![],
					headers: vec![
						hir::types::KeyValuePair {
							key: "Authorization".into(),
							type_: hir::types::Type::String,
							description: "The API Key of the User".into(),
							parameters: vec![("prefix".into(), "Api-Key".into())],
							comment: None,
						},
						hir::types::KeyValuePair {
							key: "X-Model".into(),
							type_: hir::types::Type::String,
							description: "The Model of the User".into(),
							parameters: vec![],
							comment: None,
						},
					],
					parameters: vec![],
					query: vec![],
					metadata: vec![],
					comment: None,
				}],
				headers: vec![],
				parameters: vec![],
				query: vec![],
				comment: None,
			}],
		},
	};
	let mut parsed_intermediate_representation = hir_lowering::types::Ir::default();
	input_hir.parse(
		hir_lowering::types::ParserVariables::default(),
		&mut parsed_intermediate_representation,
	);
	let mut generated_file_list = HashMap::new();
	parsed_intermediate_representation.markdown(&mut generated_file_list);
	// generated_file_list.write(Path::new("test-artifacts/test-2"));
}
