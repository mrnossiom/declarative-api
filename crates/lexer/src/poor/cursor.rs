use crate::poor::Token;
use std::str::Chars;

pub struct Cursor<'a> {
	len_remaining: usize,
	chars: Chars<'a>,

	#[cfg(debug_assertions)]
	prev: char,
}

pub(crate) const EOF_CHAR: char = '\0';

impl<'a> Cursor<'a> {
	#[must_use]
	pub fn from_source(source: &'a str) -> Cursor<'a> {
		Cursor {
			len_remaining: source.len(),
			chars: source.chars(),
			#[cfg(debug_assertions)]
			prev: EOF_CHAR,
		}
	}

	/// Returns the last eaten symbol.
	#[cfg(debug_assertions)]
	pub(super) const fn prev(&self) -> char {
		self.prev
	}

	/// Peeks the next symbol from the input stream without consuming it.
	/// If requested position doesn't exist, `EOF_CHAR` is returned.
	/// However, getting `EOF_CHAR` doesn't always mean actual end of file,
	/// it should be checked with `is_eof` method.
	pub(super) fn first(&self) -> char {
		self.chars.clone().next().unwrap_or(EOF_CHAR)
	}

	/// Peeks the second symbol from the input stream without consuming it.
	pub(super) fn _second(&self) -> char {
		let mut iter = self.chars.clone();
		iter.next();
		iter.next().unwrap_or(EOF_CHAR)
	}

	/// Checks if there is nothing more to consume.
	pub(super) fn is_eof(&self) -> bool {
		self.chars.as_str().is_empty()
	}

	/// Returns amount of already consumed symbols.
	///
	/// # Panics
	/// When used in a file that is over 4GiB
	pub(super) fn pos_within_token(&self) -> u32 {
		u32::try_from(self.len_remaining - self.chars.as_str().len())
			.expect("loaded sources can't go over 4 GiB")
	}

	/// Resets the number of bytes consumed to 0.
	pub(super) fn reset_pos_within_token(&mut self) {
		self.len_remaining = self.chars.as_str().len();
	}

	/// Moves to the next character.
	pub(super) fn bump(&mut self) -> Option<char> {
		let c = self.chars.next()?;

		#[cfg(debug_assertions)]
		{
			self.prev = c;
		}

		Some(c)
	}

	/// Eats symbols while predicate returns true or until the end of file is reached.
	pub(super) fn eat_while(&mut self, mut predicate: impl FnMut(char) -> bool) {
		while predicate(self.first()) && !self.is_eof() {
			self.bump();
		}
	}

	pub(super) fn eat_decimal_digits(&mut self) -> bool {
		let mut has_digits = false;
		loop {
			match self.first() {
				'_' => {
					self.bump();
				}
				'0'..='9' => {
					has_digits = true;
					self.bump();
				}
				_ => break,
			}
		}
		has_digits
	}
}

pub struct Iter<'a>(Cursor<'a>);

impl<'a> Iterator for Iter<'a> {
	type Item = Token;

	fn next(&mut self) -> Option<Self::Item> {
		let token = self.0.next_token();

		if token.kind.is_eof() {
			None
		} else {
			Some(token)
		}
	}
}

impl<'a> IntoIterator for Cursor<'a> {
	type Item = Token;
	type IntoIter = Iter<'a>;

	fn into_iter(self) -> Self::IntoIter {
		Iter(self)
	}
}

#[cfg(test)]
mod tests {
	use crate::tests::{ATTR, EXAMPLE, URLS};

	macro_rules! assert_tokenize {
		($variant:literal, $src:ident) => {
			paste::paste! {
				#[test]
				fn [<tokenize_ $variant>]() {
					let vec = super::Cursor::from_source($src)
						.into_iter()
						.collect::<Vec<_>>();
					insta::assert_debug_snapshot!(vec);
				}
			}
		};
	}

	assert_tokenize!("example", EXAMPLE);
	assert_tokenize!("attribute", ATTR);
	assert_tokenize!("urls", URLS);
}
