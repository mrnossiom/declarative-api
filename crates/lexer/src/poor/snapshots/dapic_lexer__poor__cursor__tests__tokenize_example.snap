---
source: crates/lexer/src/poor/cursor.rs
expression: vec
---
[
    Token {
        kind: LineComment(
            Some(
                Inner,
            ),
        ),
        length: 65,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: LineComment(
            Some(
                Inner,
            ),
        ),
        length: 36,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: LineComment(
            Some(
                Inner,
            ),
        ),
        length: 35,
    },
    Token {
        kind: Whitespace,
        length: 2,
    },
    Token {
        kind: Ident,
        length: 4,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 2,
    },
    Token {
        kind: Ident,
        length: 4,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Literal(
            Str {
                terminated: true,
            },
        ),
        length: 12,
    },
    Token {
        kind: Whitespace,
        length: 2,
    },
    Token {
        kind: Ident,
        length: 11,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Literal(
            Str {
                terminated: true,
            },
        ),
        length: 25,
    },
    Token {
        kind: Whitespace,
        length: 2,
    },
    Token {
        kind: Ident,
        length: 7,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Literal(
            Str {
                terminated: true,
            },
        ),
        length: 7,
    },
    Token {
        kind: Whitespace,
        length: 3,
    },
    Token {
        kind: Ident,
        length: 4,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBracket,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 2,
    },
    Token {
        kind: Literal(
            Str {
                terminated: true,
            },
        ),
        length: 36,
    },
    Token {
        kind: Whitespace,
        length: 2,
    },
    Token {
        kind: Literal(
            Str {
                terminated: true,
            },
        ),
        length: 44,
    },
    Token {
        kind: Whitespace,
        length: 2,
    },
    Token {
        kind: Literal(
            Str {
                terminated: true,
            },
        ),
        length: 40,
    },
    Token {
        kind: Whitespace,
        length: 2,
    },
    Token {
        kind: CloseBracket,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: CloseBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 2,
    },
    Token {
        kind: Ident,
        length: 5,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 7,
    },
    Token {
        kind: Semi,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 2,
    },
    Token {
        kind: Ident,
        length: 5,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 9,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 2,
    },
    Token {
        kind: Ident,
        length: 4,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 9,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 3,
    },
    Token {
        kind: Ident,
        length: 7,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 4,
    },
    Token {
        kind: LineComment(
            Some(
                Outer,
            ),
        ),
        length: 11,
    },
    Token {
        kind: Whitespace,
        length: 4,
    },
    Token {
        kind: LineComment(
            Some(
                Outer,
            ),
        ),
        length: 20,
    },
    Token {
        kind: Whitespace,
        length: 4,
    },
    Token {
        kind: LineComment(
            Some(
                Outer,
            ),
        ),
        length: 35,
    },
    Token {
        kind: Whitespace,
        length: 4,
    },
    Token {
        kind: Ident,
        length: 13,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 11,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Literal(
            Str {
                terminated: true,
            },
        ),
        length: 37,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Or,
        length: 1,
    },
    Token {
        kind: At,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 6,
    },
    Token {
        kind: Colon,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Literal(
            Str {
                terminated: true,
            },
        ),
        length: 9,
    },
    Token {
        kind: Or,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 4,
    },
    Token {
        kind: LineComment(
            None,
        ),
        length: 77,
    },
    Token {
        kind: Whitespace,
        length: 5,
    },
    Token {
        kind: Ident,
        length: 7,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 6,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Literal(
            Str {
                terminated: true,
            },
        ),
        length: 23,
    },
    Token {
        kind: Whitespace,
        length: 3,
    },
    Token {
        kind: CloseBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 4,
    },
    Token {
        kind: Ident,
        length: 4,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 7,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 4,
    },
    Token {
        kind: Ident,
        length: 4,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBrace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 7,
    },
    Token {
        kind: CloseBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 5,
    },
    Token {
        kind: Ident,
        length: 5,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 6,
    },
    Token {
        kind: Ident,
        length: 7,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 3,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Literal(
            Str {
                terminated: true,
            },
        ),
        length: 46,
    },
    Token {
        kind: Whitespace,
        length: 5,
    },
    Token {
        kind: CloseBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 6,
    },
    Token {
        kind: Ident,
        length: 4,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 3,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 6,
    },
    Token {
        kind: Ident,
        length: 5,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 7,
    },
    Token {
        kind: LineComment(
            Some(
                Outer,
            ),
        ),
        length: 69,
    },
    Token {
        kind: Whitespace,
        length: 7,
    },
    Token {
        kind: Ident,
        length: 4,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 6,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Literal(
            Str {
                terminated: true,
            },
        ),
        length: 12,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Or,
        length: 1,
    },
    Token {
        kind: At,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 6,
    },
    Token {
        kind: Colon,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Literal(
            Str {
                terminated: true,
            },
        ),
        length: 6,
    },
    Token {
        kind: Or,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 6,
    },
    Token {
        kind: CloseBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 7,
    },
    Token {
        kind: Ident,
        length: 4,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Literal(
            Number,
        ),
        length: 3,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 7,
    },
    Token {
        kind: Ident,
        length: 4,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 2,
    },
    Token {
        kind: Whitespace,
        length: 7,
    },
    Token {
        kind: LineComment(
            None,
        ),
        length: 9,
    },
    Token {
        kind: Whitespace,
        length: 6,
    },
    Token {
        kind: CloseBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 5,
    },
    Token {
        kind: CloseBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 4,
    },
    Token {
        kind: CloseBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 5,
    },
    Token {
        kind: Ident,
        length: 4,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 5,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 5,
    },
    Token {
        kind: Ident,
        length: 4,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 3,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 6,
    },
    Token {
        kind: Ident,
        length: 4,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Literal(
            Number,
        ),
        length: 3,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 7,
    },
    Token {
        kind: LineComment(
            None,
        ),
        length: 8,
    },
    Token {
        kind: Whitespace,
        length: 7,
    },
    Token {
        kind: LineComment(
            None,
        ),
        length: 9,
    },
    Token {
        kind: Whitespace,
        length: 7,
    },
    Token {
        kind: LineComment(
            None,
        ),
        length: 3,
    },
    Token {
        kind: Whitespace,
        length: 6,
    },
    Token {
        kind: CloseBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 7,
    },
    Token {
        kind: Ident,
        length: 4,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Literal(
            Number,
        ),
        length: 3,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 7,
    },
    Token {
        kind: LineComment(
            None,
        ),
        length: 9,
    },
    Token {
        kind: Whitespace,
        length: 6,
    },
    Token {
        kind: CloseBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 5,
    },
    Token {
        kind: CloseBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 4,
    },
    Token {
        kind: CloseBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 3,
    },
    Token {
        kind: CloseBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 2,
    },
    Token {
        kind: CloseBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 3,
    },
    Token {
        kind: Ident,
        length: 5,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 7,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 3,
    },
    Token {
        kind: Ident,
        length: 4,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 6,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Literal(
            Str {
                terminated: true,
            },
        ),
        length: 24,
    },
    Token {
        kind: Whitespace,
        length: 3,
    },
    Token {
        kind: Ident,
        length: 5,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 6,
    },
    Token {
        kind: Whitespace,
        length: 3,
    },
    Token {
        kind: Ident,
        length: 8,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 6,
    },
    Token {
        kind: Whitespace,
        length: 2,
    },
    Token {
        kind: CloseBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 3,
    },
    Token {
        kind: Ident,
        length: 4,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: Ident,
        length: 5,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 2,
    },
    Token {
        kind: CloseBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: CloseBrace,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
]
